worlds/undertale/Rules.py

```diff
@@ -0,0 +1,269 @@
+        elif area == 2:
+            return self.multiworld.get_region("Waterfall", player).entrances[0].parent_region.name
+        elif area == 3:
+            return self.multiworld.get_region("Hotland", player).entrances[0].parent_region.name
+        elif area == 4:
+            return self.multiworld.get_region("Core", player).entrances[0].parent_region.name
+
+    def _undertale_is_route(self, player: int, route: int):
+        if route == 3:
+            return (self.multiworld.route_required[player].current_key == "all_routes")
+        if self.multiworld.route_required[player].current_key == "all_routes":
+            return True
+        if route == 0:
+            return (self.multiworld.route_required[player].current_key == "neutral")
+        if route == 1:
+            return (self.multiworld.route_required[player].current_key == "pacifist")
+        if route == 2:
+            return (self.multiworld.route_required[player].current_key == "genocide")
+        return False
+
+    def _undertale_has_plot(self, player: int, item: str):
+        if item == "Goat Plush":
+            return (self.has("Goat Plush", player) or self.has("Progressive Plot",player,self._undertale_reqplot(player,"Goat Plush")))
+        elif item == "Complete Skeleton":
+            return (self.has("Complete Skeleton", player) or self.has("Progressive Plot",player,self._undertale_reqplot(player,"Complete Skeleton")))
+        elif item == "Snow Shovel":
+            return (self.has("Snow Shovel", player) or self.has("Progressive Plot",player,self._undertale_reqplot(player,"Snow Shovel")))
+        elif item == "Fish":
+            return (self.has("Fish", player) or self.has("Progressive Plot",player,self._undertale_reqplot(player,"Fish")))
+        elif item == "Heat Suit":
+            return (self.has("Heat Suit", player) or self.has("Progressive Plot",player,self._undertale_reqplot(player,"Heat Suit")))
+        elif item == "Cooking Set":
+            return (self.has("Cooking Set", player) or self._undertale_is_route(player,2) or self.has("Progressive Plot",player,self._undertale_reqplot(player,"Cooking Set")))
+        elif item == "Microphone":
+            return (self.has("Microphone", player) or self._undertale_is_route(player,2) or self.has("Progressive Plot",player,self._undertale_reqplot(player,"Microphone")))
+        elif item == "Bridge Tools":
+            return (self.has("Bridge Tools", player) or self.has("Progressive Plot",player,self._undertale_reqplot(player,"Bridge Tools")))
+        elif item == "Mettaton Plush":
+            return (self.has("Mettaton Plush", player) or self.has("Progressive Plot",player,self._undertale_reqplot(player,"Mettaton Plush")))
+        elif item == "DT Extractor":
+            return (self.has('DT Extractor', player) or self.has("Progressive Plot",player,self._undertale_reqplot(player,"DT Extractor")))
+
+    def _undertale_can_level(self, exp: int, lvl: int):
+        if (exp >= 10 and lvl == 1):
```

Comment:
Lots of redundant `()` parens in this file. Shouldn't cause an issue, but something to be aware of.